src/RF24Logger.h

```diff
@@ -0,0 +1,167 @@
+{
+private:
+   RF24LogHandler *handler;
+
+public:
+
+   /**
+    * Initializes the appender to nullptr
+    */
+   RF24Logger();
+
+   /**
+    * set the instance's handler
+    * @param handler The log handler where the messages will be redirected.
+    */
+   void setHandler(RF24LogHandler *handler);
+
+   /**
+    * ouput an ERROR message
+    * @param vendorId A scoping identity of the message's origin
+    * @param message The message to output
+    * @param args consumable arguments
+    */
+   template<class T> void error(
+         const __FlashStringHelper *vendorId, T message, ...)
+   {
+      if (handler == nullptr)
+      {
+         return;
+      }
+      va_list args;
+      va_start(args, message);
+      handler->log(RF24LogLevel::ERROR, vendorId, message, &args);
+   }
+
+   /**
+    * output a message to WARN the reader
+    * @param vendorId A scoping identity of the message's origin
+    * @param message The message to output
+    * @param args consumable arguments
+    */
+   template<class T> void warn(
+         const __FlashStringHelper *vendorId, T message, ...)
+   {
+      if (handler == nullptr)
+      {
+         return;
+      }
+      va_list args;
+      va_start(args, message);
+      handler->log(RF24LogLevel::WARN, vendorId, message, &args);
+   }
+
+   /**
+    * output an INFO message
+    * @param vendorId A scoping identity of the message's origin
+    * @param message The message to output
+    * @param args consumable arguments
+    */
+   template<class T> void info(
+         const __FlashStringHelper *vendorId, T message, ...)
+   {
+      if (handler == nullptr)
+      {
+         return;
+      }
+      va_list args;
+      va_start(args, message);
+      handler->log(RF24LogLevel::INFO, vendorId, message, &args);
+   }
+
+   /**
+    * output a message to help developers DEBUG their source code
+    * @param vendorId A scoping identity of the message's origin
+    * @param message The message to output
+    * @param args consumable arguments
+    */
+   template<class T> void debug(
+         const __FlashStringHelper *vendorId, T message, ...)
+   {
+      if (handler == nullptr)
+      {
+         return;
+      }
+      va_list args;
+      va_start(args, message);
+      handler->log(RF24LogLevel::DEBUG, vendorId, message, &args);
+   }
+
+   /**
+    * output a TRACE message that points to an event's instigator
+    * @param vendorId A scoping identity of the message's origin
+    * @param message The message to output
+    * @param args consumable arguments
+    */
+   template<class T> void trace(
+         const __FlashStringHelper *vendorId, T message, ...)
+   {
+      if (handler == nullptr)
+      {
+         return;
+      }
+      va_list args;
+      va_start(args, message);
+      handler->log(RF24LogLevel::TRACE, vendorId, message, &args);
+   }
+
+   template<class T> void log(
+            uint8_t logLevel, const __FlashStringHelper *vendorId, T message, ...)

```

Comment:
This is a nice method - now I see, thanks to @2bndy5.

With this method we can log with any log level (and sublevel). According to the compiler, every method: info(), warn(), error(), debug() and trace() takes 80B of FLASH memory. We have 5 of those methods but because of the templates every method offers two variants, so we have 10 methods in which takes 800B of FLASH. That's a lot. 
Proposition: we could just remove those error(), warn(), info(), debug(), trace() and focus just on log() and even offer a few overloaded variants of log() to support different accepted data types? What do you think?

Reply:
>Proposition: we could just remove those error(), warn(), info(), debug(), trace() and focus just on log() 

would macros be a better alternative? I'm used to having methods like info(), debug(), etc

```cpp
#ifndef info(id, msg)
#define info(id, msg) log(INFO, id, msg)
#endif
```
But I'm not sure about the scoping there. you can probably tell I haven't been trained in the rigors of C (with the exception of `printf()`)

Reply:
>you can probably tell I haven't been trained in the rigors of C

No. I'm not quite sure what the standards say about using macros. Are they good or bad? I think they will help to save some FLASH memory of the core part of the library. Those methods ```info()```, ```warn()```, ```error()```, etc. are some kind of beautifiers because the same may be achieved just with one method ```log(uint8_t logLevel,...)```. Every singe method like ```info()``` takes 256B of FLASH. Maybe it would be worth to introduce macros here.

I would name those macros with capital letters, like ```RF24LOG_INFO(vendorId, msg)``` or something similar. Capital letters could mean here that we deal with some kind of global methods (macro). At the moment I'm not so familiar with macros naming conventions.

Reply:
> Every singe method like info() takes 256B of FLASH. Maybe it would be worth to introduce macros here.

But how much memory will each invocation take when it's "inlined" with macros? We can also probably instruct the compiler to always inline the function, should it be necessary.

I think it can be kept as a function at the moment and optimization can be discussed in a separate issue later.

Reply:
>But how much memory will each invocation take when it's "inlined" with macros? We can also probably instruct the compiler to always inline the function, should it be necessary.

Hm, will the usage of macro result in inlined function invocation? I was thinking that than a macro will be replaced with a method invocation, so actually we will have only one method ```log(uint8_t logLevel,...)``` and macros will be used only as a beautifier. Ok we can discus it later in a separate issue.